[SE-3520] Adds waffle flag to enable overriding existing transcripts on import

nizarmah opened the pull request:
This pull-request is a follow up pull request, related to [this comment from a previous pull-request](https://github.com/edx/edx-val/pull/266#issuecomment-717822495).

Every time we import a new transcript, the existing and the new transcripts are compared by computing their content hash and comparing it.
This helps us identify whether the transcript uploaded is a duplicate of an already existing transcript, or if it is a different transcript that is being uploaded even though a transcript for the same video already exists.

Previously, if a video transcript for a specific video already exists, the new transcript wouldn't get uploaded. Now, if the transcript is different than the already existing one for a certain video, it does get uploaded and overrides it.

**JIRA tickets**: SE-3520, [OSPR-5117](https://openedx.atlassian.net/browse/OSPR-5117)

**Discussions**: [GitHub comment on previous pull-request](https://github.com/edx/edx-val/pull/266#issuecomment-717822495)

**Sandbox URL**:
- LMS: https://edxval-pr268.sandbox.opencraft.hosting/
- Studio: https://studio.edxval-pr268.sandbox.opencraft.hosting/

**Installation Instructions**:

- Install using:
```python
pip install --upgrade git+https://github.com/open-craft/edx-val.git@nizar/overriding_existing_non_duplicate_transcripts#egg=edxval
```

- Apply migrations to LMS
```bash
/edx/bin/edxapp-migrate-lms
```


**Testing instructions**:

- Testing using the Sandbox
    1. Login to the sandbox with `staff`/`edx`.
    1. Go to _Django Admin -> Waffle (django-waffle) -> Flag_ and activate the `edxval.override_existing_imported_transcripts` flag.
    1. Go to Studio and create a new course.
    1. Import one of the courses exported and attached below.
    1. Open the only unit in the course, edit the video, and download the transcript.
    1. Now import the other course, please.
    1. Open the only unit in the course, edit the video, and download the transcript.
    1. The transcript should be different than the other downloaded on step 4.
- Testing using Makefile
    1. Run `make requirements`
    1. Then run `make test`

> ##### Test Courses to Import
> - [First Course](https://github.com/edx/edx-val/files/5506797/course.rarp5q2h.tar.gz)
> - [Second Course](https://github.com/edx/edx-val/files/5506796/course.o5eu1225.tar.gz)

**Author notes and concerns**:

* The tox tests aren't broken due to my changes, they are broken when it comes to python 38 with django 30. I verified that the same problem is happening right now on `master`. 

**Reviewers**
- [x] @gabor-boros
- [ ] edX reviewer[s] TBD

**Settings**
```yaml
EDXAPP_EXTRA_REQUIREMENTS:
  - name: git+https://github.com/open-craft/edx-val.git@nizar/overriding_existing_non_duplicate_transcripts#egg=edxval
```

edx-webhook commented:
Thanks for the pull request, @nizarmah! I've created [OSPR-5117](https://openedx.atlassian.net/browse/OSPR-5117) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

natabene commented:
@nizarmah Thank you for your contribution. Please let me know once it is ready for our review.

gabor-boros commented:
👍  🎉 

- [x] I tested what's in the test instructions
- [x] I read through the code
- NA I checked for accessibility issues
- NA Includes documentation
- NA I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.

nizarmah commented:
@natabene this is ready for edX's review :+1: :smile:

natabene commented:
@kashifch Could you please review this?

DawoudSheraz commented on edxval/api.py:
```diff
@@ -1149,54 +1155,69 @@ def import_transcript_from_fs(edx_video_id, language_code, file_name, provider,
+    # check if transcript content already exists
+    if (
+        existing_transcript and
+        is_duplicate_file(
+            new_transcript_content_file,
+            existing_transcript.transcript_content_hash
```
How will this be backward compatible(given default value is '')?

nizarmah replied:
If the hash is empty, then the file would not be treated as a duplicate, and it would be uploaded.

Would you prefer that I apply the `generate_file_content_hash` on each existing uploaded file instead? The migration would take longer but it would be backwards compatible. 

> I'll add a commit for that once I target the other comments, that way if it isn't needed, we can just revert it.

nizarmah replied:
I have added backwards compatibility now :+1: sorry for not doing that earlier

DawoudSheraz commented on edxval/models.py:
```diff
@@ -518,15 +547,10 @@ def create_or_update(cls, video, language_code, metadata, file_data=None):
 
         transcript_name = metadata.get('file_name')
 
+        video_transcript.save()
+
```
Why this save here?

nizarmah replied:
That's a line I changed and it went unnoticed by me, sorry about that.

DawoudSheraz commented on edxval/tests/test_api.py:
```diff
@@ -1770,7 +1770,7 @@ def test_multiple_external_transcripts_for_language(self):
         # Verify transcript record is created with correct data i.e sub field transcript.
         expected_transcripts = [
-            dict(constants.VIDEO_TRANSCRIPT_CUSTOM_SRT, video_id=edx_video_id, language_code='en')
+            dict(constants.VIDEO_TRANSCRIPT_CUSTOM_SJSON, video_id=edx_video_id, language_code='en')
```
Why is this changed?

nizarmah replied:
Because the behavior changed. Here's a more detailed explanation: 

First, we need to identify that:
- `sub_transcript_file_name` is an _srt_ video transcript
- `ext_transcript_file_name` is an _sjson_ video transcript

```
# Import xml with empty edx video id.
edx_video_id = api.import_from_xml(
    etree.fromstring('<video_asset/>'),
    '',
    self.file_system,
    constants.EXPORT_IMPORT_STATIC_DIR,
    {
        'en': [sub_transcript_file_name, ext_transcript_file_name]
    }
)
```

If we take a look at the import segment of code, shown above, we can see that for the same *language_code* was given two different files/transcripts.

**Previously**, if a transcript was already uploaded/imported for a certain video, then importing that same video with a new transcript did not overwrite it. So, the initially imported `sub_transcript_file_name` transcript does not get replaced, and thus the file remains an _srt_ transcript.

**Currently**, if a transcript was already uploaded/imported for a certain video, then importing the same video with a new transcript will overwrite it if the content is different. Since the `sub_transcript_file_name` and the `ext_transcript_file_name` transcripts have different file contents, then when the `ext_transcript_file_name` transcript was uploaded, the transcript got replaced and became an _sjson_ transcript. 

Accordingly, we had to update the test :+1:

I tried to keep it brief, so please let me know if you'd like me to go into more details :slightly_smiling_face:

DawoudSheraz replied:
Thanks for the brief. I would say to update the test docstring to cover this change. It will be easier to follow in the future.

DawoudSheraz reviewed (commented):
Added a few comments. There can be some delay with this PR while some internal evaluation is done around this change. Thank You

nizarmah commented on edxval/utils.py:
```diff
@@ -243,3 +245,40 @@ def validate_generated_images(value, max_items):
+
+    Returns:
+        if file is duplicate (boolean)
+    """
```
This should be addressed now. The method takes two files instead, as you requested :+1:

nizarmah replied:
@DawoudSheraz I just realized that there won't be any need to change any migrations or model changes related to the transcript content hash, since we're computing the content hash when trying to import the transcript.

Accordingly, I added a commit which removes the field and its migrations completely.

If you'd like me to revert that change, please let me know. 

Computing the content hash every time isn't so efficient, but it's definitely more straight forward from a code point of view.

nizarmah commented:
Thanks a lot for your review @DawoudSheraz 
I've addressed 3 of the comments you left with commits, and I clarified why the test was changed :+1: 

> There can be some delay with this PR while some internal evaluation is done around this change.

No worries :slightly_smiling_face: I totally understand

Please don't hold back any requests/changes that might be needed to accept the PR, even if I have to fully rework it

DawoudSheraz commented on edxval/utils.py:
```diff
@@ -243,3 +245,41 @@ def validate_generated_images(value, max_items):
+    content_hash = hashlib.sha256()
+
+    with closing(uploaded_file.open()) as file_content:
+        for chunk in iter(lambda: file_content.read(65536), b""):
```
Out of curiosity: Is there any bytes limitation on iter? I am interested about the choice of 65536 when reading the content. How will it behave if the contents to be read have bytes less than 65536?

nizarmah replied:
> Is there any bytes limitation on iter? I am interested about the choice of 65536 when reading the content.

No, not that I know of. 65536 bytes is about 64 kilobytes. I specified that size in order to avoid allocating more memory to reading the transcript.

> How will it behave if the contents to be read have bytes less than 65536?

It would behave similarly to looping over an array with a single item. 

---

I changed this part to `file_content.read()` directly, instead, disregarding the memory constraints. 

Please let me know if you'd like me to revert that change :+1:

DawoudSheraz commented on edxval/tests/test_utils.py:
```diff
@@ -0,0 +1,55 @@
+        file_data = ContentFile(file_content)
+        other_file_data = ContentFile(file_content)
+
+        is_duplicate = is_duplicate_file(file_data, other_file_data)
```
nit: `self.assertTrue(is_duplicate_file(file_data, other_file_data))`. No need to create variable just to store the output.

DawoudSheraz commented on edxval/tests/test_utils.py:
```diff
@@ -0,0 +1,55 @@
+
+        is_duplicate = is_duplicate_file(file_data, other_file_data)
+
+        self.assertFalse(is_duplicate)
```
Same comment here.

DawoudSheraz commented on edxval/tests/test_api.py:
```diff
@@ -1785,7 +1785,7 @@ def test_external_internal_transcripts_conflict(self):
         en_transcript_file_name = 'external-transcript-en.sjson'
         utils.create_file_in_fs(
-            constants.VIDEO_TRANSCRIPT_CUSTOM_SJSON['file_data'],
+            self.transcript_data1['file_data'],
```
Same for this test case.

nizarmah replied:
By the way, the change for this test case was actually a mistake. I should've changed it differently. Accordingly, I updated the docstring for this test case, and the previous one.

In addition, I also made the necessary changes to this test case so that it imports two different transcripts, an external and an internal one. And then the internal one would be overridden by the external one, because the external one gets imported after, and its contents are different.

I would appreciate it if you could give this test another look please, when you have time @DawoudSheraz :+1:

DawoudSheraz reviewed (approved):
Overall, good to go. Some nits comment to address before merging. Please update edxval version in setup.py. Again, there might be a small delay in the PR merge and release. Thanks for your wait on the review.

DawoudSheraz commented on setup.py:
```diff
@@ -46,7 +46,7 @@ def load_requirements(*requirements_paths):
 
 
-VERSION = '1.4.3'
+VERSION = '1.4.4'
```
Please rebase your branch and update the version to 1.4.5. The latest release is 1.4.4

nizarmah replied:
Yes, sorry about that, I missed your earlier comment about this :flushed:

DawoudSheraz reviewed (approved):
Please rebase and update the branch. Once done, the PR will be good to merge.

nizarmah commented:
@DawoudSheraz I updated the version, as you requested.

Thanks a lot for your review! I hope you have a great day!

DawoudSheraz commented:
@nizarmah Hi. A follow-up question. How will this change handle `If video ID is already present in edxval then do not overwrite any video transcripts`? Because transcript is associated with Video and not CourseVideo. The changes will overwrite the transcript for the video in the original course too.

nizarmah commented:
@DawoudSheraz Hmm, so this change would always overwrite the video transcript if the video ID is already present and the transcript is different from the existing one. 
So yes, as you mentioned, this change will overwrite the transcript for all courses that use that video ID.

The reason behind this is the perspective I had which is: "Video Transcripts shouldn't be course specific, but specific to each Video. So if a Video's Transcript gets updated, it should get updated for every course that uses that Video."

However, that's something that my perspective won't be enough on, and I'd love to know if you'd like me to make any changes to what I've done, because I definitely don't have enough context to make such a decision for such a change.

Hope this answers your question, and I'm looking forward to any tweak I can do in my approach to get this accepted :+1:

DawoudSheraz commented:
@azarembok Hi. Can you provide your thoughts on this PR? Thanks

nizarmah commented:
Any updates on this @DawoudSheraz ?

DawoudSheraz commented:
@nizarmah Hello. Apologies for the delay. I have asked for another internal review. Once that approval is in, I will be merging and releasing this change. Thanks

DawoudSheraz commented:
@nizarmah Hello. Thanks for your wait on this review. There has been some internal discussion around the changes involved in this PR. Though the change is risky, it is ok to merge behind a django settings/waffle flag(your choice). The decision is that this change will not be enabled on edX but openedX community is free to enable it on their installations.

nizarmah commented:
@DawoudSheraz no problem 😃 Glad things worked out at the end. I really appreciate the decision to enable this behind a django settings/waffle flag ❤️ I'll need some time to do that change, since my plate is pretty full at the moment, but I'll let you know when the change is ready 👍🏼 

The change should be ready by January 4, approximately. Sorry for that long delay! 🙁

DawoudSheraz commented:
> @DawoudSheraz no problem 😃 Glad things worked out at the end. I really appreciate the decision to enable this behind a django settings/waffle flag ❤️ I'll need some time to do that change, since my plate is pretty full at the moment, but I'll let you know when the change is ready 👍🏼
> 
> The change should be ready by January 4, approximately. Sorry for that long delay! 🙁

No worries, take your time. Let me know when this is ready. Thank You

nizarmah commented:
@DawoudSheraz hope you're having a nice holiday :) I just wanted to let you know that I am planning to use `edx-toggles` for creating a waffle switch to enable overriding existing transcripts.

Please let me know if you'd like me to take a different approach or if you'd like me to directly just use `django-waffle` without `edx-toggles` (although the latter could be useful since it enforces namespaces to be included). 👍🏼

DawoudSheraz commented:
@nizarmah Hello. I am ok with edx-toggles or whatever method suits you best. I saw the PR in edx-toggles but it seems it is still in review. If that change is a breaking change, you can constraint edx-toggle version in edx-val for the time being.

nizarmah commented:
@DawoudSheraz Happy new year 🙂 I didn't come empty handed haha, the changes are ready for your review 👍🏼 

I also updated the sandbox, in case you'd like to test, and made `staff` superuser, so that you can enable/disable the Waffle Flag [`edxval.override_existing_imported_transcripts`](https://edxval-pr268.sandbox.opencraft.hosting/admin/waffle/flag/6/change/) however you'd like.

---

By the way, the tox tests fail for python 3.8 and django 3.0.

This applies to all branches for the time being, so it wasn't caused by my changes. I might consider submitting a pull request to handle that upgrade, during my freetime, in case some other changes are more of a priority for edX than that one 👍🏼  

---

Sorry that the changes are a lot, it's mainly due to the requirements and the tests.
I changed the tests significantly, making them data driven, that way both cases can be tested in the same method. 
Let me know please if you have any objection regarding that.

I also wanted to mention that the "exit cases" for the `import_transcript_from_fs` might be ugly, and separated, but I feel like adding those cases might be nicer than encasing everything in nested if statements .
Let me know if you'd like me to change that please.

By the way, sorry about the force push when merging master. 
I had the changes prepared in 7 neat commit messages, but part of the merge bled into the code changes, so I had to revert (until after where you have already reviewed), and then those 7 commits into one 🙁

DawoudSheraz commented on edxval/models.py:
```diff
@@ -447,6 +447,28 @@ def filename(self):
+
+        # save the transcript file
+        if file_data:
+            with closing(file_data.open()) as transcript_content:
```
Is the open(seek) call necessary here?

nizarmah replied:
This has been there since #87 ([click here to go to exact line](https://github.com/edx/edx-val/pull/87/files#diff-7bd2e8ec7f96c5cb70570aa07f67a7d44bd9c23309e6bdb8f587385c9f0830adR463-R466))

So I was hesitant to change something that was already there.

But I changed it, as you suggested, and the tests are running fine. Also, my manual testing went fine... 🤔

DawoudSheraz commented on edxval/tests/test_api.py:
```diff
@@ -1787,103 +1789,182 @@ def test_multiple_external_transcripts_different_langauges(self):
-            self.file_system,
-            constants.EXPORT_IMPORT_STATIC_DIR
-        )
+        There are two different cases here, based on whether **overriding existing transcripts is enabled**.
+
+        If overriding existing transcripts is **disabled**:
+            If a transcript was already uploaded/imported for a certain video, then importing that same video with
+            a new transcript does not overwrite it.
+            In this case:
+                1. Import `sub_transcript_file_name`
+                1. Import `ext_transcript_file_name`
+            Since both have the same video id, then `ext_transcript_file_name` would not be imported, so the transcript
+            would be the `sub_transcript_file_name`.
+
+        If overriding existing transcripts is **enabled**:
+            If a transcript was already uploaded/imported for a certain video, then importing the same video with a new
+            transcript will overwrite it, if and only if the content is different.
+            In this case,
+                1. Import `sub_transcript_file_name`
+                1. Import `ext_transcript_file_name`
+            Since both have the same video id, and `ext_transcript_file_name` has different content, it would get
+            imported, so the transcript would be the `ext_transcript_file_name`.
```
Great explanation 👍🏽

DawoudSheraz commented on edxval/config/waffle.py:
```diff
@@ -0,0 +1,28 @@
+    )
+
+
+OVERRIDE_EXISTING_IMPORTED_TRANSCRIPTS = WaffleFlag(
+    waffle_name('override_existing_imported_transcripts'),
+    module_name=__name__,
+)
```
nit: add waffle/toggle documentation as per edx-toggles specs.

DawoudSheraz reviewed (commented):
The changes are good. Thanks for updating the PR. Two questions/commits to address before merging.

nizarmah commented:
@DawoudSheraz I addressed your latest comments 👍🏼 
Let me know if there are any other changes you'd like me to make 😃 

By the way, this is one of the significant contributions I've made, so thanks for helping out with it!

DawoudSheraz reviewed (approved):
Please rebase and squash the commits. Thanks for your contribution.

nizarmah commented:
@DawoudSheraz I have rebased and squashed 👍🏼 

Thanks a lot for your review and your time! 
Have a wonderful day! 😃

edx-webhook commented:
@nizarmah 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-val&entry.867055334=https://github.com/edx/edx-val/pull/268&entry.1484655318=https://github.com/nizarmah&entry.752974735=2020-11-08+17:04&entry.1917517419=2021-01-04+13:25&entry.2133058324=Yes) so we can improve your experience in the future.

DawoudSheraz commented:
@nizarmah Thanks for your contribution, effort, patience, and cooperation on this PR. I have drafted https://github.com/edx/edx-val/tree/1.4.5 against your changes. Once it is pushed to PyPI, the changes will become part of edx-platform with the next requirements update.

nizarmah commented:
https://github.com/edx/edx-val/blob/cd54e796e0e87c5a8a4f6a4243d67b660c1642ba/edxval/config/waffle.py#L25-L38

@robrap I believe the requested documentation changes for [`edxval.override_existing_imported_transcripts`](https://docs.google.com/spreadsheets/d/1xWbEL6oNu6D84WKBs3aViLRM40xk-vmnmmGAeLyR09A/edit?ts=6008a109&pli=1#gid=0&range=A178) are included in this pull request.

Would you mind please confirming that? Or letting me know what changes would be required?

nizarmah commented:
Actually, I notice the problem now... It is:
https://github.com/edx/edx-val/blob/cd54e796e0e87c5a8a4f6a4243d67b660c1642ba/edxval/config/waffle.py#L25

I'll create a pull request to fix that 👍🏼
